Merge Minecraftian's fix to the previous merge

JasonTheKitten opened the pull request.

Minecraftian14 committed:
Fix stage one

The Overlaying component is showing.

The Animation is working good.

Minecraftian14 committed:
init Commit

Minecraftian14 committed:
bruh

Minecraftian14 committed:
Removed unrequired changes

JasonTheKitten commented on src/main/java/everyos.browser.webicitybrowser/everyos/browser/webicitybrowser/gui/component/OverlyingBlockComponent.java:
```diff
@@ -7,7 +7,7 @@
 
 public class OverlyingBlockComponent extends BlockComponent {
-    private final Component contentView;
+    public final Component contentView;
```
Except for the JavaDom source folder, we are leaning away from public variables.
There are still some cases were public and protected variables are used in the code, but we want to avoid adding more.
Keep this as private and, if absolutely necessary, create a getter method. However, it seems as if the code that you made this public for is commented out, so you probably don't need a getter.

Minecraftian14 replied:
Ah yes, I realised it very late, that it was better to use layout.

Moreover, I think I can also remove the limitation of adding in only one component.

I'll look into those today

Minecraftian14 replied:
Not fixing this issue, as a lot of classes have to be made Generic (all <Component>UI classes).
While that's not a bad thing, it's a problem for UIManager. Which only yields Component and not a possible Generic.
That can still be fixed by casting, but that's not a good idea, as we already have the casted method.

JasonTheKitten commented on src/main/java/everyos.browser.webicitybrowser/everyos/browser/webicitybrowser/gui/ui/OverlyingBlockComponentUI.java:
```diff
@@ -1,72 +1,103 @@
-        contentPaneUI = uimgr.get(contentPane, getParent());
-        contentPane.bind(contentPaneUI);
-        return new ComponentUI[]{contentPaneUI};
+        component = (OverlyingBlockComponent) c;
```
We also have a method to do this. .casted(Type.class)

Minecraftian14 replied:
Having long chaining methods is not always a good idea. But yeah, casted does it's job good.

Actually... You can make getComponent itself a generic. So you won't even have to pass in the class object.

JasonTheKitten replied:
I had that previously
https://github.com/WebicityBrowser/Webicity/blob/71d90ace7bb2c454e9f480cd3096ff33b1e6c92b/src/main/java/everyos.engine.ribbon.ui.simple/everyos/engine/ribbon/ui/simple/SimpleComponentUI.java#L83
I think it disappeared when I was refactoring

Minecraftian14 replied:
Not fixing this issue, as a lot of classes have to be made Generic (all UI classes).
While that's not a bad thing, it's a problem for UIManager. Which only yields Component and not a possible Generic.
That can still be fixed by casting, but that's not a good idea, as we already have the casted method.

JasonTheKitten commented on src/main/java/everyos.browser.webicitybrowser/everyos/browser/webicitybrowser/gui/ui/OverlyingBlockComponentUI.java:
```diff
@@ -1,72 +1,103 @@
+        }
 
-        if(component.isInvisible()) return;
+        private float progress = 0;
```
Define this before any methods (i.e. right below where we define bounds).

Minecraftian14 replied:
Sure, I sometimes like defining variables used by only one method right above it.

It's okay though I'll move it up

JasonTheKitten commented on src/main/java/everyos.browser.webicitybrowser/everyos/browser/webicitybrowser/gui/ui/OverlyingBlockComponentUI.java:
```diff
@@ -1,72 +1,103 @@
+            int h = bounds.getHeight();
 
-        r.useForeground();
+            int ha = (int) (h * progress);
```
What does ha stand for? Any chance the variable can be renamed?

Minecraftian14 replied:
😅
Okay... I can rename it to one of
heightAnimated
hAnimated 
h_anim
h_eff
hEffective
heightEffective
And so on.

JasonTheKitten replied:
heightAnimated sounds good

JasonTheKitten commented on src/main/java/everyos.browser.webicitybrowser/everyos/browser/webicitybrowser/gui/ui/OverlyingBlockComponentUI.java:
```diff
@@ -1,72 +1,103 @@
+        component = (OverlyingBlockComponent) c;
+        this.appearence = new OverlyingBlockComponentUIAppearence();
+        layout = (InlineBlockLayout) getLayout();
+        layout.setConsiderChildren(false);
```
Can we rename this to setAutoManageChildren?

Minecraftian14 replied:
Yeah we can.
In essence, when we say "consider" it also means that we will not define layouts out calculated them for children. But here we are not really doing anything with their layouts so... Okay 👌

JasonTheKitten commented on src/main/java/everyos.engine.ribbon.renderer.skijarenderer/everyos/engine/ribbon/renderer/skijarenderer/RibbonSkijaRenderer.java:
```diff
@@ -119,6 +119,11 @@ private void setClip(int x, int y, int l, int h) {
 
+	@Override
+	public void translate(int dx, int dy) {
+		this.x+=dx; this.y+=dy;
```
Why can't translate be done via the UI code?
Also, if I'm correct, this won't be clipped properly;.

Minecraftian14 replied:
Oh no it actually works just fine!

In overlaying component UI
If I draw anything above y=0, it will be drawn over webicity button!!

That's why I first clip it so that it won't draw over the button.

Later, I can easily draw the round rect above so that it will be clipped, this giving a nice animation 

But this doesn't affect children. I can set their draw positions!!!!!

That's why, when I translate the renderer up, it sets the y=0 over webicity button, but still clipping the part I wanted.

Therefore, children are also drawn nicely.


As such, clipping is working the way I expected it to!

JasonTheKitten commented on src/main/java/everyos.engine.ribbon.ui.simple/everyos/engine/ribbon/ui/simple/layout/InlineBlockLayout.java:
```diff
@@ -34,7 +34,8 @@
 	private MouseListener mouseListener;
 	private MouseListener externalMouseListener;
-	
+	private boolean considerChildren = true;
```
Can you rename this variable to autoManageChildren?

JasonTheKitten commented on src/main/java/everyos.browser.webicitybrowser/everyos/browser/webicitybrowser/util/TimeSystem.java:
```diff
@@ -0,0 +1,26 @@
+    private static long lastTime = -1;
+    private static long delta = -1;
+
+    public static void reset() {
```
Static is a no-go here. What happens if we open multiple windows?

Minecraftian14 replied:
Let it stay like that, I'll think of a solution later

JasonTheKitten replied:
Ok. Mark it as TODO at least though.

Minecraftian14 committed:
fix: removed a lot of unnecessary things from OverlyingBlockComponent.
Most probably, one can now add many components in there w/o any issues.
So the restriction of having only one component is removed.

Minecraftian14 committed:
fix: Renamed "consider" attribute for children painting and rendering to "auto manage".
Renamed "ha" to "heightAnimated" for better context relationship.
Now TimeSystem.lastTime is a static variable. At the most it can cause big (one time) delta value, but that will be sufficiently before any animations start.
